presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkDiskBasedBroadcastDependency.java

```diff
@@ -0,0 +1,117 @@
+        this.maxBroadcastSize = requireNonNull(maxBroadcastSize, "maxBroadcastSize cannot be null");
+        this.tempStorage = requireNonNull(tempStorage, "tempStorage cannot be null");
+        this.tempDataOperationContext = requireNonNull(tempDataOperationContext, "tempDataOperationContext cannot be null");
+        this.queryCompletionDeadline = queryCompletionDeadline;
+    }
+
+    @Override
+    public Broadcast<List<PrestoSparkStorageHandle>> executeBroadcast(JavaSparkContext sparkContext)
+            throws SparkException, TimeoutException
+    {
+        List<PrestoSparkStorageHandle> broadcastValue = broadcastDependency.collectAndDestroyDependenciesWithTimeout(computeNextTimeout(queryCompletionDeadline), MILLISECONDS).stream()
+                .map(Tuple2::_2)
+                .collect(toList());
+
+        long compressedBroadcastSizeInBytes = broadcastValue.stream()
+                .mapToLong(metadata -> metadata.getCompressedSizeInBytes())
+                .sum();
+        long uncompressedBroadcastSizeInBytes = broadcastValue.stream()
+                .mapToLong(metadata -> metadata.getUncompressedSizeInBytes())
+                .sum();
+
+        log.info(String.format("Got back %d pages. compressedBroadcastSizeInBytes: %d; uncompressedBroadcastSizeInBytes: %d",
+                broadcastValue.size(),
+                compressedBroadcastSizeInBytes,
+                uncompressedBroadcastSizeInBytes));
+
+        long maxBroadcastSizeInBytes = maxBroadcastSize.toBytes();
+
+        if (compressedBroadcastSizeInBytes > maxBroadcastSizeInBytes) {
+            throw exceededLocalBroadcastMemoryLimit(maxBroadcastSize, format("Compressed broadcast size: %s", succinctBytes(compressedBroadcastSizeInBytes)));
+        }
+
+        if (uncompressedBroadcastSizeInBytes > maxBroadcastSizeInBytes) {
+            throw exceededLocalBroadcastMemoryLimit(maxBroadcastSize, format("Uncompressed broadcast size: %s", succinctBytes(uncompressedBroadcastSizeInBytes)));
+        }
+
+        broadcastVariable = sparkContext.broadcast(broadcastValue);
+        return broadcastVariable;
+    }
+
+    @Override
+    public void destroy()
+    {
+        if (broadcastVariable == null) {
+            return;
+        }
+
+        try {
+            // Delete the files
+            for (PrestoSparkStorageHandle diskPage : broadcastVariable.getValue()) {
+                TempStorageHandle storageHandle = tempStorage.deserialize(diskPage.getSerializedStorageHandle());
+                tempStorage.remove(tempDataOperationContext, storageHandle);
+                log.info("Deleted broadcast spill file: " + storageHandle.toString());
+            }
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException("Unable to delete broadcast spill file", e);
+        }
+
+        // Destroy broadcast variable
+        broadcastVariable.destroy();
```

Comment:
question: Do you think the broadcast variable be destroyed first? Or should the files be deleted first?

Reply:
We cannot destroy the variable first since we use it to get the list of storage handles that needs to be deleted. Once we have the list, it does not matter which one is destroyed first.